### ISSUE TITLE
Exception in EF Core when completing a workflow.

### ISSUE THREAD

**Issue by solutionmania** [elsa-workflows/elsa-core#4526]
I have just upgraded my project to run the v3.0.0-preview.768 packages and I'm getting the exception below when attempting to start a workflow. This is just a simple workflow with a single "Write Line" activity in it.

If I comment-out the call to `runtime.UseEntityFrameworkCore()` in program.cs, the same workflow works without a problem.

```
    elsa.UseWorkflowRuntime(runtime =>
    {
        // Store detils of running workflows in SQL Server
        runtime.UseEntityFrameworkCore(ef =>
        {
            ef.UseSqlServer("Server=localhost;Database=ElsaPlayground;Integrated Security=True;TrustServerCertificate=True;");
        });
    });
```

After upgrading the packages, I deleted all database tables to allow the migrations to create them from scratch.

**Exception:**
```
Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while saving the entity changes. See the inner exception for details.
 ---> Microsoft.Data.SqlClient.SqlException (0x80131904): Conversion failed when converting the nvarchar value 'Completed' to data type int.
   at Microsoft.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__209_0(Task`1 result)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
ClientConnectionId:8c403381-ab79-4b5d-a9b3-62a93b05cad1
Error Number:245,State:1,Class:16
   --- End of inner exception stack trace ---
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.SqlServer.Update.Internal.SqlServerModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(IList`1 entriesToSave, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(StateManager stateManager, Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
   at Elsa.EntityFrameworkCore.Extensions.QueryableExtensions.BulkUpsertAsync[TDbContext,TEntity](TDbContext dbContext, IList`1 entities, Expression`1 keySelector, CancellationToken cancellationToken)
   at Elsa.EntityFrameworkCore.Common.Store`2.SaveManyAsync(IEnumerable`1 entities, Expression`1 keySelector, Func`4 onSaving, CancellationToken cancellationToken)
   at Elsa.EntityFrameworkCore.Common.Store`2.SaveManyAsync(IEnumerable`1 entities, Expression`1 keySelector, Func`4 onSaving, CancellationToken cancellationToken)
   at Elsa.EntityFrameworkCore.Common.EntityStore`2.SaveManyAsync(IEnumerable`1 entities, Func`4 onSaving, CancellationToken cancellationToken)
   at Elsa.EntityFrameworkCore.Modules.Runtime.EFCoreActivityExecutionStore.SaveManyAsync(IEnumerable`1 records, CancellationToken cancellationToken)
   at Elsa.Workflows.Runtime.Middleware.Workflows.PersistActivityExecutionLogMiddleware.InvokeAsync(WorkflowExecutionContext context)
   at Elsa.Workflows.Runtime.Middleware.Workflows.PersistBookmarkMiddleware.InvokeAsync(WorkflowExecutionContext context)
   at Elsa.Workflows.Runtime.Middleware.Workflows.PersistentVariablesMiddleware.InvokeAsync(WorkflowExecutionContext context)
   at Elsa.Workflows.Core.Pipelines.WorkflowExecution.WorkflowExecutionPipeline.ExecuteAsync(WorkflowExecutionContext context)
   at Elsa.Workflows.Core.Services.WorkflowRunner.RunAsync(WorkflowExecutionContext workflowExecutionContext)
   at Elsa.Workflows.Core.Services.WorkflowRunner.RunAsync(Workflow workflow, RunWorkflowOptions options, CancellationToken cancellationToken)
   at Elsa.Workflows.Runtime.Services.WorkflowHost.StartWorkflowAsync(StartWorkflowHostOptions options, CancellationToken cancellationToken)
   at Elsa.Workflows.Runtime.Services.DefaultWorkflowRuntime.StartWorkflowAsync(IWorkflowHost workflowHost, StartWorkflowRuntimeOptions options)
   at Elsa.Workflows.Runtime.Services.DefaultWorkflowRuntime.StartWorkflowAsync(IWorkflowHost workflowHost, StartWorkflowRuntimeOptions options)
   at Elsa.Workflows.Runtime.Services.DefaultWorkflowRuntime.StartWorkflowAsync(String definitionId, StartWorkflowRuntimeOptions options)
   at Elsa.Workflows.Api.Endpoints.WorkflowDefinitions.Execute.Execute.HandleAsync(Request request, CancellationToken cancellationToken)
   at FastEndpoints.Endpoint`2.ExecAsync(CancellationToken ct)
   at FastEndpoints.Endpoint`2.ExecAsync(CancellationToken ct)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Elsa.Http.Middleware.WorkflowsMiddleware.InvokeAsync(HttpContext httpContext)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)

```

**Comment by solutionmania** [comment]
Possibly related to #4520

**Comment by solutionmania** [comment]
Appears to be fixed in v3.0.0-preview.768 - thanks!

**Comment by sfmskywalker** [comment]
That’s good news! But it also makes me wonder what the heck changed that fixed it 😅

**Comment by solutionmania** [comment]
Sorry - my mistake (long day at work) - I just deleted my comment - it wasn't fixed after all!

**Comment by sfmskywalker** [comment]
No worries, I’m actually glad we can get to the bottom of this now :)

Would it be possible for you to attach a (stripped down) version of your project that I can try locally to reproduce the issue? It would make it easier to troubleshoot if I can see the error firsthand.

**Comment by sfmskywalker** [comment]
Hi @solutionmania ,

Thanks for the repro!

When I run the project, it throws an error that is not able to find a table. Which makes sense, since I didn't run the migrations. Migrations, by default, are executed from a startup task from an IHostedService implementation - which are only executed from host applications (Generic Host or Web Host) and not from Console apps - unless you do so explicitly yourself.

When I do run the migrations, the application starts and finishes with no issue.
Here's the complete Program.cs file:

```csharp
using Elsa.EntityFrameworkCore.Extensions;
using Elsa.EntityFrameworkCore.Modules.Runtime;
using Elsa.Extensions;
using Elsa.Workflows.Core.Activities;
using Elsa.Workflows.Core.Contracts;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.DependencyInjection;

// Setup service container.
var services = new ServiceCollection();

// Add Elsa services to the container.
services.AddElsa(elsa =>
{
    elsa.UseWorkflowRuntime(runtime =>
    {
        // Store details of running workflows in SQL Server
        runtime.UseEntityFrameworkCore(ef =>
        {
            //ef.UseSqlServer("Server=localhost;Database=ElsaPlayground;Integrated Security=True;TrustServerCertificate=True;");
            ef.UseSqlite();
        });
    });
});

// Build the service container.
var serviceProvider = services.BuildServiceProvider();

// Run migrations.
await RunMigrationsAsync<RuntimeElsaDbContext>();

// Define a simple workflow that writes a message to the console.
var workflow = new Sequence
{
    Activities =
    {
        new WriteLine("Hello World!"),
        new WriteLine("Goodbye cruel world...")
    }
};

// Resolve a workflow runner to execute the workflow.
var workflowRunner = serviceProvider.GetRequiredService<IWorkflowRunner>();

// Execute the workflow.
await workflowRunner.RunAsync(workflow);
return;

async Task RunMigrationsAsync<TDbContext>() where TDbContext : DbContext
{
    var dbContextFactory = serviceProvider.GetRequiredService<IDbContextFactory<TDbContext>>();
    await using var dbContext = await dbContextFactory.CreateDbContextAsync();
    await dbContext.Database.MigrateAsync();
    await dbContext.DisposeAsync();
}
```

I used SQLite instead of SQL Server, since I am not running on a Windows machine. But, that shouldn't make any difference.

**Comment by solutionmania** [comment]
Hi,

I just created a brand new database, then re-ran my repro project with your additions to run the migrations and I'm still seeing exactly the same issue.

SqlException: Conversion failed when converting the nvarchar value 'Completed' to data type int.

Does SQLite use the same migrations as SQL Server?

I haven't dug into the source to find out what field it's trying to write to - but making an educated guess - the 'ActivityExecutionRecords' table that's been created in my new DB has a field called 'Status' which is an 'int' - so if that's where it's trying to store the value 'Completed' - that'd be the issue!

**Comment by solutionmania** [comment]
I can also confirm that I **don't** see the problem with SQLite - but it is 100% repeatable with SQL Server.

I'll try and use my repro project against the Elsa source later and see if I can work out what's going on.

**Comment by sfmskywalker** [comment]
> Does SQLite use the same migrations as SQL Server?

Actually, we generate specialized migrations for each provider, so it's possible that we forgot to update them for SQL Server (despite the fact that we use a script to update all providers) or that there's an issue specific with this provider's migrations.

I'll do more testing with SQL Azure for which SQL Server migrations will be used.

> the 'ActivityExecutionRecords' table that's been created in my new DB has a field called 'Status' which is an 'int' - so if that's where it's trying to store the value 'Completed' - that'd be the issue!

That's almost 100% certainly the issue - there's a custom converter that configures EF Core to map enums to strings rather than ints - but, if the generated schema expects an int still, then that explains perfectly what you've been seeing.

I'll get it fixed soon. Thanks for the details 👍🏻

**Comment by sfmskywalker** [comment]
After generating migrations, it updated the generated code. So we must have forgotten to run the script after having changed the EF Core configuration.

**Comment by solutionmania** [comment]
Excellent - thanks for looking into this!

**Comment by sfmskywalker** [comment]
No worries! I have just pushed newly generated migrations that should fix the issue.

**Comment by solutionmania** [comment]
Can confirm this is fixed in 3.0.0-preview.785. Many thanks!
